packages/hooks/src/Provider.tsx

```diff
@@ -0,0 +1,94 @@
+import { JsonRpcSigner } from '@ethersproject/providers/src.ts/json-rpc-provider';
+import React from 'react';
+import { ethers } from 'ethers';
+import WalletConnectProvider from '@walletconnect/web3-provider';
+import Web3Modal from 'web3modal';
+
+export interface Web3ContextType {
+  connectWallet: (() => void) | undefined;
+  signer: JsonRpcSigner | null | undefined;
+  userAddress: string | null | undefined;
+  disconnectWallet: (() => void) | undefined;
+  network: string;
+  chainId: number | undefined;
+  connected: boolean;
+  provider: ethers.providers.Web3Provider | null | undefined;
+}
+
+export const Web3Context = React.createContext<Web3ContextType | undefined>(undefined);
+
+export interface ProviderProps {
+  /**
+   * @dev The network you want to connect to.
+   * @example 'mainnet'
+   * @type string
+   */
+  network: string;
+  /**
+   * @dev Your Infura project ID. This is required if you want to support WalletConnect.
+   * @type string
+   */
+  infuraId?: string;
+}
+
+/**
+ * @dev The global provider that handles and stores all the web3 connections. Wrap your entire App with this component.
+ * @param children Your app.
+ * @param network The network you want to connect to.
+ * @param infuraId Your Infura project ID. This is required if you want to support WalletConnect.
+ */
+export const Provider: React.FC<ProviderProps> = ({ children, network, infuraId }) => {
+  const [signer, setSigner] = React.useState<null | JsonRpcSigner>(null);
+  const [provider, setProvider] = React.useState<null | ethers.providers.Web3Provider>(null);
+  const [userAddress, setUserAddress] = React.useState<null | string>(null);
+  const [web3Modal, setWeb3Modal] = React.useState<null | Web3Modal>();
```

Comment:
what's the difference between null and undefined here? if this is only ever undefined or `Web3Modal`, then lets remove the `null` type